flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/input/MultiStateKeyIteratorTest.java

```diff
@@ -125,4 +185,117 @@ public void testIteratorRemovesFromAllDescriptors() throws Exception {
                             .count());
         }
     }
+
+    /** Test for lazy enumeration of inner iterators. */
+    @Test
+    public void testIteratorPullsSingleKeyFromAllDescriptors() throws AssertionError {
+        CountingManyKeysKeyedStateBackend keyedStateBackend =
+                createManyKeysKeyedStateBackend(100_000_000);
+        MultiStateKeyIterator<Integer> testedIterator =
+                new MultiStateKeyIterator<>(descriptors, keyedStateBackend);
+
+        boolean hasnext = testedIterator.hasNext();
+
+        Assert.assertEquals(
+                "Unexpected number of keys enumerated",
+                1,
+                keyedStateBackend.numberOfKeysEnumerated);
+    }
+
+    /**
+     * Mockup {@link AbstractKeyedStateBackend} that counts how many keys are enumerated.
+     *
+     * <p>Generates a configured nmber of integer keys, only method actually implemented is {@link
+     * CountingManyKeysKeyedStateBackend#getKeys(java.lang.String, java.lang.Object)}
+     */
+    static class CountingManyKeysKeyedStateBackend extends AbstractKeyedStateBackend<Integer> {
```

Comment:
How about renaming to `CountingKeyNumberKeyedStateBackend` or just `CountingKeysKeyedStateBackend` ?